Add helpers for raw method creation

B-head opened the pull request:
## Feature
Easily write code for manual bindings. The error propagation operator can now be used for error handling.
Let's compare this with the [original code](https://github.com/godot-rust/godot-rust/blob/c69c5a0241a2206ec45e2a8028922ca10648bcfa/gdnative-core/src/export/method.rs#L538-L591).
```rust
unsafe extern "C" fn raw_calc(
    this: *mut sys::godot_object,
    _method_data: *mut libc::c_void,
    user_data: *mut libc::c_void,
    num_args: libc::c_int,
    args: *mut *mut sys::godot_variant,
) -> sys::godot_variant {
    (|| {
        let _this = unsafe { Ref::<Reference>::try_from_sys(this)?.assume_safe_unchecked() };
        let user_data = unsafe { cast_sys_user_data::<RawMethod>(user_data)? };
        let mut args = unsafe { Varargs::from_sys(num_args, args) };

        let a = args.next().and_then(|v| v.to::<i64>()).unwrap_or(3);
        let b = args.next().and_then(|v| v.to::<i64>()).unwrap_or(4);

        let result =
            user_data.map(|ud| gdnative::export::catch_unwind(move || a * b + ud.inner));

        Ok::<_, Box<dyn Error>>(result??.to_variant())
    })()
    .unwrap_or_else(|err| {
        log::error(godot_site!(raw_calc), err);
        Variant::nil()
    })
    .leak()
}
```

## List of APIs to add
Most APIs are added to simplify error handling.

### `Ref::try_from_sys()`
Convert from a raw pointer, incrementing the reference counter if reference-counted.
If `obj` is null, an error is returned.

### `cast_sys_user_data()`
The user_data wrapper can be obtained directly.

Takes an opaque pointer produced by `into_user_data` and "clones" it to produce the
original instance, increasing the reference count.
If `user_data` is null, an error is returned.

### `catch_unwind()`
Wrapper function for `std:panic:catch_unwind()` that returns `PanickedError` when panicked.
`PanickedError` displays the contents if the payload passed during panic has type `&str` or `String`.

### `RawMethodBuilder`
Builder type used to register a raw method on a `NativeClass`.

## Compatibility
Since this PR only adds an API, there are no compatibility issues.

Bromeon commented:
Thank you for this pull request! 🙂 

Maybe before I have a detailed look, could you describe a bit the context and motivation? Where do you need raw methods? Why not the high-level API?

Note also that we removed an entire API around [`ClassBuilder::add_method()`](https://docs.rs/gdnative/0.9.3/gdnative/nativescript/struct.ClassBuilder.html#method.add_method) in 0.10.0, see details in https://github.com/godot-rust/godot-rust/pull/828. We specifically did so to direct the user towards the safe and easier-to-use equivalent `ClassBuilder::method()`.

B-head commented:
> Where do you need raw methods?

My use case is optimization of the wrapper generated by the `#[methods]` macro. 

The current macro generates wrappers that uses dynamic dispatching, which limits optimization by the Rust compiler. Generating wrappers that uses static dispatch is as simple as generating the direct wrappers. So I added the `RawMethodBuilder`.

Also, some may write user-defined macros for the following reasons.

- Create wrappers using method_data.
- `Rc<T>` directly to the engine so that references can be shared with Rust code. 
  - Right now there is no way to access the `Arc<T>` used in the `UserData` wrapper.
  - If I were developing a game with lots of engine interaction, I might have done this.
- To do something necessary for asynchronous processing.

> Why not the high-level API?

This is due to technical limitations. I don't know how to optimize it other than by creating direct wrappers. :slightly_frowning_face:

> Note also that we removed an entire API around [`ClassBuilder::add_method()`](https://docs.rs/gdnative/0.9.3/gdnative/nativescript/struct.ClassBuilder.html#method.add_method) in 0.10.0, see details in #828. We specifically did so to direct the user towards the safe and easier-to-use equivalent `ClassBuilder::method()`.

I would not recommend removing the API without an alternative. :sweat_smile:

I think manual binding is mostly used to circumvent the technical limitations of the high-level API. So it is reasonable to provide low-level APIs as ways of manual binding. Well, the side effect is that its API becomes unsafe. :innocent:

B-head commented:
Changed `from_sys()` to properly bounded lifetime. This is a bit of strange definition. :roll_eyes:

Bromeon commented:
bors try

B-head commented:
Reports whether `RawMethodBuilder` methods are marked as unsafe or not.

### `method_ptr` and `free_func`
These two methods are considered safe because they cannot be passed invalid function pointers.

Of course, it is possible that a method passed may be unsafe, but it will be marked as unsafe inside the implementation.

### `method_data`
This method is considered safe because the passed value is not accessible to Godot engine. It can be accessed by methods registered with `method_ptr` and `free_func`, but responsibility for safe handling rests with the registered method.

The passed value must be dropped by a method registered with `free_func` or leaks of memory will occur, but [that is considered safe in Rust](https://doc.rust-lang.org/nomicon/what-unsafe-does.html).

### `with_rpc_mode`
This is obvious. :monocle_face:

B-head commented:
> The current macro generates wrappers that uses dynamic dispatching, which limits optimization by the Rust compiler. Generating wrappers that uses static dispatch is as simple as generating the direct wrappers. So I added the RawMethodBuilder.

This was static dispatch since the `dyn` keyword was not used. Wrong! :dizzy_face:

> Create wrappers using method_data.

It seems that if data are included in a structure that implemented `Method` trait, they can be substituted for `method_data`.

```rust
struct Foo {
    bar: i64 // included data
}

impl Method<Baz> for Foo{
    fn call(&self, _this: TInstance<Baz>, args: Varargs) -> Variant {
        self.bar // use data
    }
}
```
